### ISSUE TITLE
Lock with blocking_timeout and context manager trap

### ISSUE THREAD

**Issue by gjcarneiro** [andymccurdy/redis-py#621]
Consider the following code:

``` python
with redis_client.lock("foo", blocking_timeout=2):
    # do stuff
```

What happens if the lock is already acquired by someone else for a long time and we expire the 2 second blocking timeout?  Well, redispy happily proceeds to execute the "do stuff" code without holding the lock!  This is dangerous.  I had a nasty bug in my code because of this.

Why this happens?  Because the `redis_client.lock("foo", blocking_timeout=2)` constructs a Lock object, which does `self.blocking_timeout = blocking_timeout` in the constructor.  When we `__enter__` the context manager, this code is executed:

``` python
    def __enter__(self):
        # force blocking, as otherwise the user would have to check whether
        # the lock was actually acquired or not.
        self.acquire(blocking=True)
        return self
```

Clearly the author was aware of a similar danger involving blocking=False.  However, following the code, we get to:

``` python
    def acquire(self, blocking=None, blocking_timeout=None):
        [...]
        if blocking_timeout is None:
            blocking_timeout = self.blocking_timeout
```

So, in the end, the acquire method runs with a blocking_timeout=2 parameter, which means it may return before acquiring the lock, returning False instead of True in this case.

In this case, I would expect `__enter__` to check the return value of self.acquire() and raise an exception if it is True.  I worked around it in my code, but others may fall into the same trap.  Please fix.

**Comment by omenar** [comment]
How did you worked it around in your code?

**Comment by gjcarneiro** [comment]
I made my own context manager without the bug.  Basically, check the return value of acquire and raise an exception if the lock couldn't be acquired.  Something like (simplified):

```python

class AlreadyLockedError(Exception): pass

class MyLock:

    def __init__(self, redis_client, key):
         self.lock = redis_client.lock(key, blocking_timeout=2)

    def __enter__(self):
        if self.lock.acquire():
             return self
        else:
             raise AlreadyLockedError

     def __exit__(self, type, value, tb):
         self.lock.release()

# example:
with MyLock("keyname"):
    do_something()
```

If the acquiring the lock times out, you receive an exception, instead of executing the do_something() code anyway.  Much safer this way.

**Comment by nathan-hoad** [comment]
I've just discovered this, and I've done a similar thing to @gjcarneiro. Any reason why this hasn't been fixed yet? It makes using locks as context managers quite broken as it stands now.

**Comment by jclark360** [comment]
Yeah this is nasty, if you can't get the lock an exception should be raised.

**Comment by evowise** [comment]
I had the exact same problem, took some while to pinpoint it down.

**Comment by ngfeldman** [comment]
My team ran into this too.

**Comment by andymccurdy** [comment]
Fixed in redis-py 3.0
